crates/lox-derive/src/lib.rs

```diff
@@ -0,0 +1,724 @@
+                                Ok(item) => items.push(item),
+                                Err(crate::ndm::kvn::KvnDeserializerErr::UnexpectedKeyword { .. }) |
+                                Err(crate::ndm::kvn::KvnDeserializerErr::UnexpectedEndOfInput { .. }) => break,
+                                Err(e) => Err(e)?,
+                            }
+                        }
+
+                        items
+                    }
+                });
+            }
+        }
+    }
+
+    Err(
+        syn::Error::new_spanned(field, "Malformed type for `#[derive(KvnDeserialize)]`")
+            .into_compile_error()
+            .into(),
+    )
+}
+
+fn is_value_unit_struct(item: &DeriveInput) -> bool {
+    for attr in item.attrs.iter() {
+        if attr.path().is_ident("kvn")
+            && attr
+                .parse_nested_meta(|meta| {
+                    if meta.path.is_ident("value_unit_struct") {
+                        Ok(())
+                    } else {
+                        Err(meta.error("unsupported attribute"))
+                    }
+                })
+                .is_ok()
+        {
+            return true;
+        }
+    }
+
+    false
+}
+
+fn extract_type_path(ty: &syn::Type) -> Option<&syn::Path> {
+    match *ty {
+        syn::Type::Path(ref typepath) if typepath.qself.is_none() => Some(&typepath.path),
+        _ => None,
+    }
+}
+
+fn handle_version_field(
+    type_name: &proc_macro2::Ident,
+    field: &syn::Field,
+) -> Result<proc_macro2::TokenStream, proc_macro2::TokenStream> {
+    let string_type_name = type_name.to_string();
+
+    if !string_type_name.ends_with("Type") {
+        return Err(syn::Error::new_spanned(
+            type_name,
+            "Types with \"version\" field should be of the form SomethingType",
+        )
+        .into_compile_error());
+    }
+
+    let message_type_name = string_type_name
+        .trim_end_matches("Type")
+        .to_string()
+        .to_uppercase();
+
+    let field_name = field.ident.as_ref().unwrap();
+
+    // 7.4.4 Keywords must be uppercase and must not contain blanks
+    let expected_kvn_name = format!("CCSDS_{}_VERS", message_type_name);
+
+    // Unwrap is okay because we expect this span to come from the source code
+    let field_type = extract_type_path(&field.ty)
+        .unwrap()
+        .span()
+        .source_text()
+        .unwrap();
+    let field_type_new = extract_type_path(&field.ty).unwrap();
+
+    let parser = generate_call_to_deserializer_for_kvn_type(
+        &field_type,
+        field_type_new,
+        &expected_kvn_name,
+    )?;
+
+    Ok(quote! {
+        #field_name: #parser?,
+    })
+}
+
+fn deserializer_for_struct_with_named_fields(
+    type_name: &proc_macro2::Ident,
+    fields: &syn::punctuated::Punctuated<syn::Field, syn::token::Comma>,
+    is_value_unit_struct: bool,
+) -> proc_macro2::TokenStream {
+    if &type_name.to_string() == "UserDefinedType" {
```

Comment:
Rather than converting `type_name` to string for each call, you could compare it with a static `Ident` like `Ident::new("UserDefinedType", Span::def_site())`.

Reply:
A general clean-up of the proc macro is planned after the merge.

Reply:
Ok, as it clearly works as-is, give me a shout when the cleanup PR is done and I'll review that too.